odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1671 @@
+        else:
+            # It should not be possible to get here
+            raise TypeError('comparison with scalar {} failed'.format(scalar))
+
+        OperatorRightScalarMult.__init__(self, operator=func, scalar=scalar)
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluate
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+        return (self.scalar * self.operator.gradient *
+                ScalingOperator(self.domain, self.scalar))
+
+    @property
+    def conjugate_functional(self):
+        """Convex conjugate functional of functional with scaled argument."""
+        return ConvexConjugateArgScaling(self.operator.conjugate_functional,
+                                         self.scalar)
+
+    def proximal(self, sigma=1.0):
+        """Return the proximal operator of the functional with scaled argument.
+
+        Parameters
+        ----------
+        sigma : positive float, optional
+            Regularization parameter of the proximal operator.
+
+        Returns
+        -------
+        out : `Operator`
+            Domain and range equal to domain of functional.
+        """
+        sigma = float(sigma)
+        return(proximal_arg_scaling(
+            self.operator.proximal, self.scalar))(sigma)
+
+    def derivative(self, point):
+        """Returns the derivative operator in the given point.
+
+        This function returns the linear operator
+
+            ``x --> <x, grad_f(point)>``,
+
+        where ``grad_f(point)`` is the gradient of the functional in the point
+        ``point``.
+
+        Parameters
+        ----------
+        point : `LinearSpaceVector`
+            The point in which the gradient is evaluated.
+
+        Returns
+        -------
+        out : `DerivativeOperator`
+            The linear operator that maps ``x --> <x, grad_f(point)>``.
+        """
+        return self.scalar * self.operator.derivative(self.scalar * point)
+
+
+class FunctionalComp(Functional, OperatorComp):
+
+    """Composition of a functional with an operator.
+
+    Given a functional ``func`` and an operator ``op``, such that the range of
+    the operator is equal to the domain of the functional, this corresponds to
+    the functional
+
+        ``(func * op)(x) = func(op(x))``.
+    """
+
+    def __init__(self, func, op, tmp1=None, tmp2=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        func : `Functional`
+            The left ("outer") operator
+        op : `Operator`
+            The right ("inner") operator. Its range must coincide with the
+            domain of ``func``.
+        tmp1 : `element` of the range of ``op``, optional
```

Comment:
Skip adding these for now. Makes testing harder and noone ever seems to use them anyway.